N-dimensional arrays, tensor and functions in tensor fields.

AunShiLord opened the pull request:
Good day!
As the part of our diploma paper, we have been working on this humble package. We started development when sympy/tensor was almost empty. 
Our goal was to add missing tensor class and functions to work with tensors.
Unlike your existing abstract tensor, we represent tensor as N-dimensional array with specific index character. 1 for contravariant and -1 for covariant.

To make module independent from third party libraries (ndarray of NumPy) we create class Arraypy - multi-dimensional array. Arraypy allows creating arrays of any rank. Addition and subtraction implemented. Moreover, Arraypy support arrays with custom indexes. For example, 3x3 matrix could start from index [1, 1] and end with index [4, 4]. 
Our Tensor class inherited from Arraypy. Arraypy serves as base for data, while Tensor has additional property index character.
At this moment, we have implemented tensor product and tensor contraction. All functions in other modules was done with help of Arraypy/Tensor.

Module tensor_methods also contains functions to return symmetric or asymmetric form of Tensor/Arraypy.

Module tensor_fields contains functions for working with the tensor fields:
- the calculation of the differential and the gradient of the function;
- curl and divergence of a vector field;
- the calculation of the derivative of Li and the external differentiation
  of differential forms.

All functions can be operated with input arguments specified as a Tensor. 
Some of the input arguments of functions can be a list, matrix or array of Arraypy.
The starting index of tensor may not be equal to 0. In such case, the object, which returns the function, will have the starting index corresponding to the conditions of the call.
In addition, lie_w function and dw involves working with multidimensional arrays.

Module riemannian_geometry contains functions for working with tensor fields:
- the calculation of the scalar product;
- the Christoffel symbols of the first and second kind;
- the covariant derivative of the curvature tensor;
- the Ricci tensor;
- scalar and sectional curvature.

All functions take arguments, the types of which may be such as list, matrix, or array Arraypy/Tensor. Some functions have optional parameter indicating the type of the function result. Starting index of arguments with type Arraypy or Tensor is not necessarily and by default equal to 0. The function determines the range of the index in array to return the object with the same range of index.

P.S. This is our second try. Well, last time we were not ready to share our work. But now we are REALY ready to cooperate if you are interested in any of our module. Maybe some modules will be more usefull in other sections or else... 
Anyway, Thanks.

AunShiLord committed:
Added files arraypy, tensor_fields, tensor_methods, Riemannian_Geometry

-Module "arraypy" describes Tensor and it's bases - Multidimentional arrays.
-Module "tensor_fields" add several fuctions, that is let user perform actions on tensors like finding “symmetric” tensor
-Module “Riemannian_geometry” describes functions from Riemannian geometry that can be performed on tensors.
-Module tensor_fields describes some functions on tensor fields.

AunShiLord committed:
files __init__.py, arraypy.py, tensor_methods.py updated according to pep8

Signed-off-by: AunShiLord <aunshilord@yahoo.com>

KsenijaM committed:
Renamed function

Soroche committed:
autopep8

Soroche committed:
2_autopep8

Soroche committed:
3_autopep8

KsenijaM committed:
1_autopep8

KsenijaM committed:
Renamed functions

AunShiLord committed:
Changed fucntion names in __init__ and tensor_methods

AunShiLord committed:
Renamed class arraypy -> Arraypy
in __init__, arraypy, tensot_methods, tensor_field, riemannian_geometry

AunShiLord committed:
Renamed tensor -> Tensor
in __init__, arraypy, tensor_methods, tensor_fields, riemannian_geometry

AunShiLord committed:
Renamed method and variable names:
Next_index -> next_index
Reshape -> reshape
To_matrix -> to_matrix
To_tensor -> to_tensor

AunShiLord committed:
Edited docstrings and fix with import

AunShiLord committed:
Added iterator.
Now Arraypy can be used in "for" loops:
a = Arraypy (shape)
for i in a:
    print(i)

AunShiLord committed:
fixed importError

Soroche committed:
pep8_docformatter

AunShiLord committed:
Added method __eq__ to Arraypy and Tensor

AunShiLord committed:
tensor_methods now use sympy factorial
__str__ now leaves two spaces instead of one

AunShiLord committed:
renamed methods. Updated comments
To_list -> to_list
Contract -> contract
To_arraypy -> to_arraypy

Soroche committed:
doc

Soroche committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
little_test

AunShiLord committed:
fixed issue with ind_char returning list type

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

AunShiLord committed:
Added tests for arraypy module

Soroche committed:
fg

Soroche committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
all

AunShiLord committed:
Fixed issue with .contract of 2-dimensional arrays.
Added tests for arraypy and tensor_methods.

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

KsenijaM committed:
Edited documentation and solved problem with function count

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
опечатка

AunShiLord committed:
Arraypy inherited from Basic

Soroche committed:
скопировано описание модуля

Soroche committed:
all

Soroche committed:
all2

KsenijaM committed:
комментарий на транслите

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
фы

KsenijaM committed:
all

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

KsenijaM committed:
Added tests

KsenijaM committed:
Edited documentation

Soroche committed:
change documentation

Soroche committed:
the module description

KsenijaM committed:
doc

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
rename function

AunShiLord committed:
Code corrected by autopep8.

KsenijaM committed:
description

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

KsenijaM committed:
autopep8

AunShiLord committed:
added fucntion wedge to tensor_methods.py
removed Tensor.__mul__() from arraypy.py to tensorProduct() of tensor_methods.py
Added method __mul__() to class Arraypy in arraypy.py.

All methods/functions not yet tested

Soroche committed:
change import from arraypy in tensor_fields

KsenijaM committed:
Change import from arraypy and syntax

KsenijaM committed:
change docs, corrected errors from travis-ci

Soroche committed:
change div to diverg

Soroche committed:
in test chache ddiv to diverg

AunShiLord committed:
Added functions is_symmetric, is_asymemtric to tensor_methods.py
Fucntion change_basis work in progress.
tensorProduct -> tensor_product

AunShiLord committed:
Added functions lower_index, raise_index to tensor_methods.py
Added property iter_index to arraypy.py

AunShiLord committed:
Little changes on __next__, __iter__ and .iter_index
Now it works correctly.

KsenijaM committed:
Change docs

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Conflicts:
	sympy/tensor/arraypy.py

Soroche committed:
change div to diverg in init.py

KsenijaM committed:
deleted docs line

Soroche committed:
fixed bag

Soroche committed:
fixed bag

KsenijaM committed:
fixed bag

KsenijaM committed:
fix bag

KsenijaM committed:
fixed bag

AunShiLord committed:
fixed bag

AunShiLord committed:
fixed few more bugs

AunShiLord committed:
fixed bag
Added tests to test_args.py

AunShiLord committed:
Fixed bug
temporarily removed examples from iter_index and __next__()

KsenijaM committed:
fixed bag

KsenijaM committed:
last bag

KsenijaM committed:
fixed bag

Upabjojr commented:
Wow, I just noticed this PR. This work looks amazing.

I one note: the `Tensor` class already exists in SymPy, what about renaming it something slightly different, for example `TensorArray` (this makes it clear that it's both a tensor and an array)?

AunShiLord commented:
Yay! Thanks for the answer! We've been waiting so long :)

Sure, there is still much work to do. Renaming classes and functions is not a problem. We just hope that this work will be usefull.
Please send us your comments.

AunShiLord committed:
Added .index_list to Arraypy.
Returns list of tuples - all possible indecies sorted in ascending oreder.

Upabjojr commented:
Yes, there used to be two or three people interested in tensors for SymPy in the past, but they have become inactive now. As you notice there are around 250 open pull requests, the reviewing process is currently very slow.

Upabjojr commented on sympy/core/tests/test_args.py:
```diff
@@ -2994,6 +2994,15 @@ def test_sympy__simplify__hyperexpand__G_Function():
     from sympy.simplify.hyperexpand import G_Function
     assert _test_args(G_Function([2], [1], [], []))
 
+@XFAIL
```
I think it's important to get rid of this `@XFAIL`. I will be off for the 1st of May holidays, next week I'll have a better look.

AunShiLord replied:
Sure, thanks! That was a temprorary matter to pass Travis CI tests.

KsenijaM committed:
fixed bag

AunShiLord committed:
Some works on change_basis.
WIP

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
fixed bug autopep8
add check skew-symmetric in dw, lie_w
delete 'print' in test_tensor_fields

Soroche committed:
autopep8, docformatter in tensor_methods

AunShiLord committed:
Added change basis.
Not properly tested yet.

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Conflicts:
	sympy/tensor/tensor_methods.py

AunShiLord committed:
Improved raise_index and lower_index.
Now many indicies can be raised (or lowered) at the same time.

AunShiLord committed:
Fixed bug

Upabjojr commented:
Hey!

The first thing I noticed is the `Basic` dependence of `arraypy`. The point is, `Basic` has methods for pattern matching, term rewriting, and some expression manipulation. It assumes that the objects that inherit it are immutable.

Have a look at the difference between _MutableMatrix_ and _ImmutableMatrix_ in SymPy:

```
In [1]: m1 = MutableMatrix([[1,2],[3,4]])

In [2]: m1
Out[2]: 
⎡1  2⎤
⎢    ⎥
⎣3  4⎦

In [3]: type(m1)
Out[3]: sympy.matrices.dense.MutableDenseMatrix

In [4]: isinstance(m1, Basic)
Out[4]: False

In [5]: m2 = Immut
ImmutableDenseMatrix   ImmutableMatrix        ImmutableSparseMatrix

In [5]: m2 = ImmutableMatrix(m1)

In [6]: type(m2)
Out[6]: sympy.matrices.immutable.ImmutableMatrix

In [8]: isinstance(m2, Basic)
Out[8]: True

```

As you can see, there immutable matrices inherit _Basic_, while mutable ones don't.

It makes sense that objects inheriting _Basic_ are immutable, because _Basic_ contains methods for term rewriting, that _rebuild_ the object, i.e. they take the object's _args_ and recall the constructor, all mutable elements get lost and your _arraypy_ object would get all of its components wiped off after term rewriting.

I would suggest to separate _arraypy_ in a mutable and immutable classes inheriting an abstract class, as matrices in SymPy are.

Upabjojr commented:
This example shows that an immutable matrix records all of its components in `.args` (i.e. the arguments passed to the `Basic` constructor), while the mutable matrix does not have `.args` (since it does not inherit `Basic`):
```
In [9]: m2.args
Out[9]: (2, 2, (1, 2, 3, 4))

In [10]: m1.args
AttributeError: MutableDenseMatrix has no attribute args.
```

This property should be satisfied by all classes inheriting `Basic`, as `ImmutableMatrix` does:
```
In [11]: type(m2)(*m2.args) == m2
Out[11]: True
```
which is the same as
```
In [12]: ImmutableMatrix(*m2.args) == m2
Out[12]: True
```

AunShiLord commented:
Hello!
I've been researching Matrix class inheritance hierarchy and well... It's kinda compliated. :)

..........Basic............................................Object...........
.............|................................................../.................
.............|.........................................MatrixBase............
........MatrixExpr.............................../.........|.................
.................|..................DenceMatrix..........|.................
.................|.................../..............|...........|.................
..............ImmutabeMatrix.............MutableDenceMatrix...
..................................................................................

And I dont found a Matrix class - does it even exists?

And looking at all this I don't know what to do with our Arraypy class. As for me "immutable" means that I can't change arrays elements since I set them. And I don't see a point in doing so with multidimensional arrays, which is used only in calculations.
Where am I wrong? )

toolforger commented:
(Nitpick: Put ASCII art between two lines of \`\`\` so it gets a monospaced font, then mail recipients in ASCII mode will render that in a decipherable fashion as well.)

`grep -r "class Matrix" did not give me a definition of such a class, so it probably does not exist.
Which isn't that surprising from an system architect's perspective, matrices exist in many variants (as you just found out), and the likelihood of common contracts, a common interface or even just common code decreases as the width and depth of any inheritance hierarchy increases.

Upabjojr commented:
> And I dont found a Matrix class - does it even exists?

`Matrix( )` is a shortcut to create a `MutableDenseMatrix`.

> And looking at all this I don't know what to do with our Arraypy class. As for me "immutable" means that I can't change arrays elements since I set them. And I don't see a point in doing so with multidimensional arrays, which is used only in calculations.

If you inherit `Basic`, it's better that the object be immutable, because otherwise there may be some serious bugs by using the term rewriting system. An advantage of immutability would be to include your class in SymPy expressions, and better interaction with other modules. Anyway, this is not important.

If you want your object to be mutable, I would suggest to inherit `object` instead of `Basic`.

Upabjojr commented:
Another suggestion: arraypy seems to have a complicated constructor, what about a [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) ?

Example:
```
ab = ArraypyBuilder()
ab.set_sparse(False)
ab.set_start( ... )
ab.set_dimensions( ... ... )
a = ab.build()

# `a` is the instance of arraypy created with such properties
```

AunShiLord commented:
I'd prefer to inherit `object` for now. It will be a propper aim to make it immutable later :)

I do agree that constructor is kinda complicated. The idea was to make it looks like Tensors in Maple. Maybe builder will be a better solution. It is correct but big.
Or maybe functions-constructors like `list2arraypy`?

toolforger commented:
> I'd prefer to inherit `object` for now. It will be a propper aim to make it immutable later :)

Adding immutability as an afterthought will not work very well.

> I do agree that constructor is kinda complicated. The idea was to
> make it looks like Tensors in Maple. Maybe builder will be a better
> solution.

Complicated constructors are less of a problem in Python than in most 
other languages. You do
   foo(flagblah = True, expr = someExpr, dimensions = 5)
instead of something like
   new foo().setBlah(true).setExpr(someExpr).setDimensions(5).build()

The thing to keep in mind is that in Python, parameter names are part of 
the official API and can't be changed without going through a 
deprecation cycle, so these names need to be as carefully considered as 
function names.

Upabjojr commented:
> Adding immutability as an afterthought will not work very well.

Why not? A simple idea I had, is to do something like this:

```
class ImmutableArraypy(Arraypy, Basic):

    def __init__(args ... , new_arg_with_array_elements):
        ... add array initialization then call Basic.__init__

    def __setitem__(self, key, item):
        raise error("immutable")
```

> Complicated constructors are less of a problem in Python than in most other languages. You do
>    foo(flagblah = True, expr = someExpr, dimensions = 5)
> instead of something like
>    new foo().setBlah(true).setExpr(someExpr).setDimensions(5).build()

Yes, you are right.

toolforger commented:
Am 08.05.2015 um 16:02 schrieb Francesco Bonazzi:
>> Adding immutability as an afterthought will not work very well.
>
> Why not?

Immutability has a profound effect on API decisions.
E.g. you can copy immutable objects by copying around references, with 
mutable objects, you have to look very carefully how deep you want your 
copy. So you add APIs to deal with that situation.
Caching immutable objects is easy, caching mutable objects is 
complicated (read: can cause subtle bugs unless everybody agrees about 
what can mutate and what cannot).
You cannot use mutable objects as dict keys.

 > A simple idea I had, is to do something like this:
>
> ```
> class ImmutableArraypy(Arraypy, Basic):
>
>      def __init__(args ... , new_arg_with_array_elements):
>          ... add array initialization then call Basic.__init__
>
>      def __setitem__(self, key, item):
>          raise error("immutable")
> ```

Things get very ugly if mutable objects become items inside a 
supposedly-immutable data structure. This can cause the immutable data 
structure to be effectively mutable anyway.

In an immutable framework like SymPy, all mutable classes need to have 
big red warning labels.
I.e. if you want to have a mutable and an immutable `Arraypy`, name them 
`Arraypy` and `MutableArraypy`. Unless `Arraypy` is in a subsection 
that's generally well-understood to deal with mutable data structures 
exclusively, that is.

AunShiLord commented:
Well let's summ it up. I don't realy want to make `Arraypy` immutable, cause it's already mutable and a lot of code was writen. But if SymPy want it I can do the job.:smile:
Just let me clearly know what to improve.
-Inherit `Arraypy` from `object` (if We decide to do it mutable only)
-Redisign constructor. How? Make it builder-like or make more clear parameters?

AunShiLord commented:
The little problem is that i'd like that `Arraypy` can be created in different ways.
For example
```
a = Arraypy((3,3))
```
Will create 3 on 3 matrix with indicies starting from (0,0) and ending on (2,2).

```
b = Arraypy([(0, 2), (1, 3)]
```
Will also create 3 on 3 matrix but with indicies from (0,1) to (2, 3).

```
c = Arraypy([2, 3, 1])
```
Will create 3 on 3 matrix with indicies from (1, 1) to (3, 3)

I'd prefer to leave this functionality somehow.

Upabjojr commented:
Idea 1: What about staticmethods? I.e. leave the constructor unchanged, then add static methods such as `Arraypy.create_by_shape( ... )`, `Arraypy.create_by_params( ... )`

Idea 2: refactor a bit the constructor and make it mandatory to specify the parameters, i.e. `Arraypy(shape= ... )`, `Arraypy(indexed_shape= ... )`, `Arraypy(some other param = ... )`.

Upabjojr commented:
__Idea 3__ : do not modify the constructor, and leave it as it is now, but leave a warning in the documentation, explaining that the constructor may change in the future.

toolforger commented:
> I don't realy want to make `Arraypy` immutable, cause it's already mutable and a lot of code was writen.

"Lot of code was written" clearly means we should keep `Arraypy` as it 
is. It would be nice to know how much "lot" actually is: Just a handful 
of classes in SymPy? Entire subsystems? More importantly: How widespread 
is its use in user code?

 > But if SymPy want it I can do the job.:smile:

Do you have experience with this kind of large-scale refactoring?
I'm not asking because I want to shoot your offer down (it's much 
appreciated actually); I'm after finding out what options it gives us.

One way to keep `Arraypy` at its current API would be to set up a new 
class hierarchy:

```
class Arraypy2(object)
     # Immutable base class.
     # The name should be improved :-)
class MutableArraypy2(Arraypy2)
     # Adds mutators.
class Arraypy(MutableArrayPy2)
     pass
     # Just an alias for MutableArraypy2 that old code will use.
     # This class could be deprecated, but with this overall
     # structure we can leave it in without affecting us much.
```

(BTW I'm just tossing around options. People with more knowledge about 
Arraypy's use cases will have to decide which of these to apply.)

AunShiLord commented:
> It would be nice to know how much "lot" actually is: Just a handful
of classes in SymPy? Entire subsystems? More importantly: How widespread
is its use in user code?

Not that much. But all works in modules (which is developed with my collaborators) tensor_methods.py, riemannian_geometry.py and tensor_fields.py based on `Arraypy` and `Tensor`. 

> Do you have experience with this kind of large-scale refactoring?

No, I don't. This is why it scares me a bit :)

> (BTW I'm just tossing around options. People with more knowledge about
Arraypy's use cases will have to decide which of these to apply.)

I understand. So I need a decision what to do exactly. If it would depend on me only I would leave it as it is.

> class Arraypy2(object)
     # Immutable base class.

Ain't it should inherit from Basic?

Upabjojr commented:
Well, I think we can leave the issue of immutability for the future. This code is already great.

Did you think of an alternative name for _Tensor_ ? As I wrote a while ago, there is already another class called _Tensor_ in _sympy_ (in the same module, by the way), and I suggest to avoid naming conflicts. My suggestion is to rename it _TensorArray_ .

Upabjojr commented:
There are immediate points of application for your _Tensor_ object, e.g. in [the differential geometry module](http://docs.sympy.org/latest/modules/diffgeom.html#sympy.diffgeom.metric_to_Riemann_components) a Riemann tensor is currently stored in a tuple of tuples of tuples. Otherwise, in _sympy.tensor.tensor.py_ there are external dependencies on _numpy_ which could be done away with.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1159 @@
+- the Ricci tensor;
+- scalar and sectional curvature.
+
+To implement the functions used modules: matrices and tensor
```
I would write: _The modules used to implement these functions_

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1159 @@
+and by default equal to 0. The function determines the range of the index
+in array to return the object with the same range of index.
+
+Functions are work with multidimensional arrays arraypy and tensors,
```
_Functions work_

toolforger commented:
>> Do you have experience with this kind of large-scale refactoring?
>
> No, I don't. This is why it scares me a bit :)

If it's just a simple rename, that's a good way to get one's feet wet 
with that kind of work.
That said, let's wait and see what people with more knowledge about the 
use cases think.

>> class Arraypy2(object)
>       # Immutable base class.
>
> Ain't it should inherit from Basic?

Good catch, you're right.
Actually I think that's the primary purpose of refactoring this: to get 
the array stuff into the standard SymPy type hierarchy, so we can reuse 
standard mechanisms where they make sense.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1159 @@
+    >>> Y = [3, 4]
+
+    g is a metric tensor must be symmetric matrix, arraypy or tensor
+    with valence of indices (-1, -1):
```
Why does it have to have valence (-1, -1) ? I think it can be converted to that valence, can't it?

AunShiLord replied:
You mean by lowering the indices of metric tensor?

Upabjojr commented:
I notice that the functions in _riemannian\_geometry.py_ have a lots of _if statements_ to check the types. Furthermore, they seem to be the same repeated code. What about a function converting the arguments to the right type, to be called by all those functions?

Upabjojr commented:
Still, I am looking at the code in _riemannian\_geometry.py_ , there are often many nested for-loops to calculate contractions. Don't you have a function to tell which components to contract without having to write so many for-loops?

Upabjojr commented on sympy/tensor/tensor_methods.py:
```diff
@@ -0,0 +1,477 @@
+    >>> a = list2arraypy(list(range(9)), (3,3))
+    >>> b = asymmetric(a)
+    >>> print (b)
+    0  -1.00000000000000  -2.00000000000000
```
Why are they floats?

AunShiLord replied:
There is a division in formula
`res_arr[tuple(index)] /= factorial(in_arr.rank)`
So resulted element can't be int.
But this is looking ugly, agree.

Upabjojr replied:
OK, so maybe arraypy should sympify all of its elements. That is, when assigning elements to the array, do `x = sympify(x)` before storing them.

AunShiLord replied:
Ok!

Upabjojr commented on sympy/tensor/tensor_methods.py:
```diff
@@ -0,0 +1,477 @@
+    return res_arr
+
+
+def tensor_product(first_tensor, second_tensor):
```
should this be a method of _Tensor_?

AunShiLord replied:
I can't remember why, but our scientific adviser decided to implement all tensor algebra in the separate module tensor_methods. No other reasons.
Should I turn them into tensor methods?

Upabjojr replied:
We need multiple dispatch in SymPy, which is currently unsupported. This would solve this problem.

Maybe leave a function which calls a method of _Tensor_?

Upabjojr replied:
All these methods are definitely associated with a class, and class-programming uses methods. There are many advantages, like TAB-completion and so on. I would prefer to have a function, but Python does not support generic functions.

Upabjojr replied:
By the way, this is OK, it works. Let's not focus on such subtleties. Functions are OK.

Upabjojr commented:
Lots of functions could be methods of _Tensor_, is there a reason why you'd keep them as functions?

Upabjojr commented:
Maybe something similar to [numpy.tensordot](http://docs.scipy.org/doc/numpy/reference/generated/numpy.tensordot.html) could be useful:

AunShiLord commented:
> Did you think of an alternative name for Tensor ?

Yes, I did. We were busy about inheritance questions. If we will make different versions of Arraypy, then there should be a mutable and immutable tensors.
About the name, I thought about something like `NumericTensor`. `TensorArray` is nice too.

> That said, let's wait and see what people with more knowledge about the
use cases think.

Okay, let's wait.

AunShiLord commented:
> Maybe something similar to numpy.tensordot could be useful.

Is this some special tensor product?

AunShiLord commented:
> I notice that the functions in riemannian_geometry.py have a lots of if statements to check the types. Furthermore, they seem to be the same repeated code. What about a function converting the arguments to the right type, to be called by all those functions?

We decided that user must do all preparation work by himself. This is why valency, shape, etc must be set up right.

Upabjojr commented:
> About the name, I thought about something like `NumericTensor`. `TensorArray` is nice too.

_Numeric_ sounds a bit like numbers only, don't you want to store symbols inside?

>> That said, let's wait and see what people with more knowledge about the use cases think.

> Okay, let's wait.

Who are these people?

>> Maybe something similar to numpy.tensordot could be useful.

> Is this some special tensor product?

It's a tensor product, but you then decide which axes to sum over. Have a look at [how I used it](https://github.com/sympy/sympy/blob/ac8c984166d07eb37f1814d9ebf6b9c5ab6f16d3/sympy/tensor/tensor.py#L1020) in _sympy.tensor.tensor.py_, which is also a candidate tensor module, operating at higher level. It relies on _numpy_ for array storage.

Another alternative, something similar to [numpy.trace](http://docs.scipy.org/doc/numpy/reference/generated/numpy.trace.html), which is a generalization of the matrix trace. In a tensor of rank greater than two, you have to specify which axes to sum over. I [used it here](https://github.com/sympy/sympy/blob/ac8c984166d07eb37f1814d9ebf6b9c5ab6f16d3/sympy/tensor/tensor.py#L1122).

I'm showing these functions, because they would allow to do all the contraction stuff without any need of many nested if-loops.

> We decided that user must do all preparation work by himself. This is why valency, shape, etc must be set up right.

To me, this looks like making things more complicated for end-users. That is, once you have a metric it's straightforward to change valence of an index, maybe by creating a simple function to do that. Think about a possible future end user using your code, he's got the metric with the wrong signature in IPython, puts it into your function and gets this boring error...

AunShiLord commented:
> Numeric sounds a bit like numbers only, don't you want to store symbols inside?

Agreed... Symbols can be stored, so you're right, let it be `TensorArray`

> Another alternative, something similar to numpy.trace, which is a generalization of the matrix trace.

I will figure out about tensordot. Trace over 2 axes are already implemented as `Tensor` method .contract (you would ask "Why a method here and not function". I don't know :D).

> Who are these people?

If I understand correctly, then I don't know :)

Upabjojr commented on sympy/tensor/tensor_fields.py:
```diff
@@ -0,0 +1,797 @@
+
+    >>> f=x1**2*x2 + sin(x2*x3 - x2)
+
+    args it’s a list of symbol arguments of the function f. It can be in list,
```
what kind of character is this `’`. Vim gives me unicode representation _e2 80 99_, while ordinary `'` is _27_.

Soroche committed:
change doc.text

AunShiLord committed:
Tensor -> TensorArray
Few fixes.
Also autopep8-ed.

Soroche committed:
change Tensor to TensorArray in tensor_fields(all)

Upabjojr commented:
I thought a bit overnight about the methods vs functions issue. It makes sense to use functions instead of methods, if you want to consider SymPy symbols/expressions as rank-0 tensors. That is, a method can be used only on your objects (i.e. `TensorArray.method_name(...)` ), while a function may get as parameters both a `TensorArray` instance or simple objects such as _Symbol, Add, Pow, Mul, Number_.

Upabjojr commented:
The advantage of _numpy.tensordot_ is that contraction is performed while calculating the tensor product. It is equivalent to a tensor product followed by contractions, obviously if you do those two operations separately, you'll need to create an intermediate (large) tensor, then contract it.

AunShiLord committed:
Fixed issue when change_basis and raise_index raises error if TensorArray index started not from 0.

AunShiLord committed:
Reverting bad commit

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

AunShiLord committed:
Cleaning repo from bad commit
(wtf happen?!)

KsenijaM committed:
Change Tensot to TensorArray in riemannian_geometry

Soroche committed:
add functions
add helper_functions
change init

Soroche committed:
Tensor->TensorArray in test_args
change doc in hodge_star, codiff, int_product
change doc in helper_functions
rename permutations_sign to sign_permutations

Soroche committed:
doc in tensor_fields, helper_functions

Soroche committed:
the same

AunShiLord committed:
Fixed issue with raise_index
WIP reworking .contract

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
the same

Soroche committed:
the same

AunShiLord committed:
Improved tensor product

AunShiLord committed:
lost return statment

Upabjojr commented on sympy/tensor/arraypy.py:
```diff
@@ -0,0 +1,1518 @@
+
+        """
+        if isinstance(index, int):
+            index = (index,)
```
I would put a 
```
value = sympify(value)
```
here.

Upabjojr commented:
`git checkout origin/master sympy/core/tests/test_args.py`

this will reset _test\_args.py_ to the latest version from master.

toolforger commented:
Some new thoughts:
1) To clarify: I've been thinking that Arraypy was present in SymPy, but now I see it's in the new contribution. So please ignore all my remarks about existing use cases or people who might know these, I was just barking up the wrong tree.
2) To wrap up some meandering discussion: SymPy's normal mode of operation is with subtypes of `Basic`, which are all supposed to be immutable. There are some mutable classes around for those situations where immutability didn't work, or where the author just thought it wouldn't work.
3) Stay away from multiple dispatch if you can. It is an inherently antimodular technique; if you cannot avoid it, you need to coordinate the new code with all other types that may go into a multi-dispatch function. (This is complicated by the fact that multiple dispatch has not been used for serious in SymPy yet, so you'd be one of the first explorers.)

AunShiLord committed:
Simpify.
Arraypy inherited from object again

AunShiLord committed:
reseting test_args.py

AunShiLord committed:
Changed example output.
Commented few line (they might be unnecessary)

Soroche committed:
int_product -> inner_product
add some tests
fix error in hodge_star(+codiff)

KsenijaM committed:
add function

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

AunShiLord committed:
Eq issue fix

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
fix bug in codiff

AunShiLord committed:
Reworked .contract
Fixed issue, when contract could not be executed with ranged indexies

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

AunShiLord committed:
hot fix for Arraypy sympify

AunShiLord committed:
hot fix
to allow Arraypy be an element of Arraypy

AunShiLord committed:
improvements of array arithmetics

KsenijaM committed:
Fixed bugs and added functions

AunShiLord committed:
simplify asymmetric and symmetric form

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

KsenijaM committed:
fixed bugs

Soroche committed:
delete import from helper_functions
change example in lie_w

KsenijaM committed:
fixed bugs in tests

KsenijaM committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
change example in lie_w

KsenijaM committed:
change k_sigma_li

AunShiLord committed:
-arraypy.py - fixed issue with arithmetics,
fixed issue with reshape
-tensor_methds.py - clean with autopep8
+ added new tests

AunShiLord committed:
hot fix

AunShiLord committed:
added missing import

AunShiLord committed:
fixes for python 2.x

KsenijaM committed:
fixed bugs

AunShiLord committed:
change in wedge docs

AunShiLord committed:
Merge branch 'master' of https://github.com/AunShiLord/sympy

Soroche committed:
change WedgeArray

Soroche committed:
rename Wedge_array to WedgeArray

Soroche commented:
Hello!
What about functions in other modules?
For example, tensor_fields or riemannian_geomerty?

Soroche committed:
autopep8 test_tensor_fields
add '>>>' in wedge(tensor_methods)

Upabjojr commented on sympy/tensor/arraypy.py:
```diff
@@ -0,0 +1,1414 @@
+                if index in self._output:
+                    return self._output[index]
+                else:
+                    return 0
```
`return S.Zero`

Return a _SymPy_ zero, not a Python zero.

On the header
`from sympy.core import S`

Upabjojr replied:
Suppose you have an Arraypy _a_, _a_ is sparse and _a[3]_ is zero.

`(1+a[3])/2` should be `1/2` (SymPy object), not `0.5` as with Python numbers.

Upabjojr commented on sympy/tensor/arraypy.py:
```diff
@@ -0,0 +1,1414 @@
+from sympy import Symbol
+from sympy.matrices import Matrix, MatrixSymbol
+from copy import copy
+from sympy import sympify
```
`from sympy.core.compatibility import string_types, range`
This makes sure that `range` behaves the same way in Python 2 and Python 3.

Furthermore, strings in Python 3 are slightly different. SymPy offers `string_types` to solve compatibility issues between Python 2 and 3.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1857 @@
+            for j in indices:
+                for s in indices:
+                    for k in indices:
+                        gamma[p, i, j] = 0.5 * (C[p, i, j] + g[j, s] * C[s, k, i] * g_inv[
```
Instead of _0.5_ use `S.Half`

Upabjojr commented on sympy/tensor/tests/test_riemannian_geometry.py:
```diff
@@ -0,0 +1,673 @@
+    g[1, 1] = 1
+
+    res_arr = Arraypy([4, 2, 0])
+    res_arr[0, 0, 0, 0] = -0.25 * sin(x2)**2 * cos(x2)**2
```
`S.One/4` on `/4` at the end.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1857 @@
+    The curvature tensor:
+    >>> r_li = riemann_li(C, g, var, 'a')
+    >>> print(r_li)
+    -0.25*sin(x2)**2*cos(x2)**2  0
```
`-sin(x2)**2*cos(x2)**2/4  0`

Upabjojr commented:
I put some comments about the use of floating point numbers. I think that we should avoid them in _SymPy_, as _SymPy_ has a good ability to handle fractions of integers.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1857 @@
+                                                           k],
+                                                         var[j - idx_start]) - diff(g[i,
+                                                                                      j],
+                                                                                    var[k - idx_start])) / 2
```
What about functions performing tensor products and contractions instead of nested for-loops? That would be easier to read.

Upabjojr commented on sympy/tensor/riemannian_geometry.py:
```diff
@@ -0,0 +1,1857 @@
+
+def covar_der(X, g, var, type_output='t'):
+    """Return the covariant derivative the vector field.
+    nabla X[i,j] = diff(X[j],x[i])+Sum_{k}(Gamma2[k,i,j]*X[k])
```
LaTeX is enabled in the documentation. You can define the string as `r""" ... ` to avoid the character `\` being escaped.

Upabjojr replied:
Use character \` around LaTeX expressions:
\`\nabla X_{i j} = diff(X_j,x_i)+\sum_{k}(\Gamma 2_{k i j}\cdot X_k)\`

Upabjojr commented:
Concerning the name abbreviations, do we really need them? Like `covar_der` instead of `covariant_derivative`?

A lot of users use IPython with TAB completion to immediately fill long names. I think long names are OK.

Upabjojr commented:
@bjodah @jcrist @moorepants if you have time, have a look at this PR. It would be interesting to have an N-dimensional array support in SymPy. I'm feeling like this PR got forgotten, which is a pity, given the amount of work these people have given on it.

I think `Arraypy` and `TensorArray` could be merged into a subfolder of _sympy/tensor_ (say, for example, _sympy/tensor/arrays_).

Maybe rename _Arraypy_ to _MutableNDimArray_ or _MutableNDArray_ . Maybe the same for _TensorArray_ (i.e. _MutableTensorArray_), in case someone will introduce their immutable equivalents in the future (as with matrices).

Another possibility involves separating _MutableDenseNDimArray_ and _MutableSparseNDimArray_, that is use two different types.

I think the stuff relating to differential geometry should rather be moved to `sympy.diffgeom`.

asmeurer commented:
How much work is needed to finish this? If the original author(s) haven't responded, perhaps someone else should pick it up and finish the work in a new pull request.

jksuom commented:
The size of this PR is daunting. Perhaps someone could break it into smaller pieces and feed them as separate PRs. @Upabjojr ?

bjodah commented:
I'm afraid I lack both the time and expertise needed to review this constructively.
Isn't this big enough to be a separate package? (at least initially before API is stabilized)

Upabjojr commented:
This PR appears like a jewel out of the blue. It is enormous, but if properly merged, would provide N-dimensional arrays and component-based tensor support to SymPy.

I would split it like this:
* PR with N-dim arrays
* PR with `TensorArray` (that is, an N-dim array with marked valence).
* move the rest to _sympy.diffgeom_ (maybe leave it pending, sectional curvature can wait, no urgent need for it).
* rewrite the existing code of _sympy.diffgeom_ to use _TensorArray_.

I can try to do it, maybe by the end of this year.

I think the array stuff could be put into a new directory _sympy.tensor.arrays_.

This would also close the long-sought issue of unifying all tensor/Nd-array code drafts scattered through SymPy into one module.

I mentioned this PR in the main discussion because it would solve some of the problems encountered during the discussion.

moorepants commented:
@AunShiLord we'd love to get this into sympy but it's going to take some more work to reconfigure and make it manageable. Are you able to help make this happen?
